text/3064-bitfield.md

```diff
@@ -0,0 +1,401 @@
+
+    On `x86_64`, for `N = 1`, the correct Rust type is `[u8; 2]`. For `N = 3`,
+    the correct Rust type is `u16`. In either case,  the struct requires an
+    additional `repr(align(4))` annotation
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This feature allows you to write C-compatible structs by applying the `bitfield`
+attribute to fields:
+
+```rust
+#[repr(C)]
+struct X {
+    #[bitfield(1)] a: u8,
+    #[bitfield(8)] b: u8,
+    #[bitfield(0)] _: u8,
+    #[bitfield(2)] d: u8,
+}
+```
+
+corresponds to the C struct
+
+```c
+struct X {
+    unsigned char a: 1;
+    unsigned char b: 8;
+    unsigned char  : 0;
+    unsigned char d: 2;
+};
+```
+
+As you can see in the example, unnamed bit-fields in C are written with the
+reserved identifier `_` in Rust. When you have a struct with an unnamed
+bit-field, you cannot access that field. On the other hand, you also do not have
+to define a value for that field when constructing the struct:
+
+```rust
+let x = X { a: 1, b: 1, d: 1 };
+let X { a, b, d } = x;
+```
+
+Just like in C, you cannot take the address of a bit-field:
+
+```rust
+fn f(x: &X) -> &u8 {
+    &x.a // !!! does not compile
+}
+```
+
+The value of the `N` in `bitfield(N)` must be a non-negative integer literal.
+This literal must not be larger than the bit-size of the type of the field.
+
+In debug builds, when you write to a bit-field, Rust performs overflow checks.
+If the value to be stored does not fit in the bit-field, the overflow check
+fails and the operation panics:
+
+```rust
+fn f(x: &mut X) {
+    x.a = 6; // !!! overflow check fails
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The struct [grammar][struct-grammar] is modified:
+
+[struct-grammar]: https://doc.rust-lang.org/reference/items/structs.html
+
+```diff
+    | TupleStruct
+ 
+ StructStruct :
+-   struct IDENTIFIER  GenericParams? WhereClause? ( { StructFields? } | ; )
++   struct IDENTIFIER  GenericParams? WhereClause? ( { StructBody? } | ; )
+ 
+ TupleStruct :
+    struct IDENTIFIER  GenericParams? ( TupleFields? ) WhereClause? ;
+ 
+-StructFields :
+-   StructField (, StructField)* ,?
++StructBody :
++   StructBodyElement (, StructBodyElement)* ,?
++
++StructBodyElement :
++     StructField
++   | UnnamedStructElement
+ 
+ StructField :
+    OuterAttribute*
+    Visibility?
+    IDENTIFIER : Type
+ 
++UnnamedStructElement :
++   OuterAttribute*
++   Visibility?
++   `_` : Type
++
+ TupleFields :
+    TupleField (, TupleField)* ,?
+```
+
+UnnamedStructElements look syntactically like StructFields except that their
+names must be `_` which cannot be the name of a StructField. Semantically, they
+are not fields of the struct they are contained in.
+
+UnnamedStructElements are used when determining the layout of the struct but are
+otherwise ignored. Since they are not fields, they cannot be accessed, do not
+appear in the construction of a struct, etc.
```

Comment:
```suggestion
UnnamedStructElements are used when determining the layout of the struct but are
otherwise ignored. Since they are not fields, they cannot be accessed, do not
appear in the construction of a struct, etc. However, the contents of UnnamedStructElements
are unspecified, allowing reading from and transmuting to/from bitfield structs without incurring
undefined behavior.
```

Reply:
Assume that structs could have associated types:

```
struct X {
	type MyType = i32;
	field1: u8,
	field2: u16,
}
```

Does the statement

>However, the contents of AssociatedTypes are unspecified, allowing reading from and transmuting to/from structs with associated types without incurring undefined behavior.

make sense to you?

Reply:
Of course not, because types don't hold values.

Reply:
AssociatedType is not a type. It's a non-terminal in the Rust grammar. It's a part of the syntax, and syntax does not exist at runtime. As such, they don't have contents except in the syntax sense.

Reply:
If you agree they have no content, then I don't understand what you're getting at.

Reply:
If you agree that non-terminals of the Rust grammar have no content, then you must also agree that UnnamedStructElement has no content.

Reply:
Associated types don't exist at runtime, they're type system constructs. The same isn't true for padding bits, since ***padding bits exist at runtime***. These aren't some abstract things that are compiled away, they're very real and very in-existence parts of data that just aren't named variants of the bitfield as a whole

Reply:
>Associated types don't exist at runtime, they're type system constructs.

Exactly. The same statement holds for UnnamedStructElements. They are type system constructs whose only purpose is to affect the size and alignment of the struct and the position of the fields.

I think you are still confused about this because they look like fields to make specifying these type system effects easier for the user.

Reply:
Padding bits exist at runtime, that's just a fact. They're bits that *pad* the surrounding struct to the correct size and/or alignment, as you said. These bits have a value of *some* kind at runtime, and that value needs to be defined in some way, be it undefined value, unspecified value or a specified value. An unspecified value is the best choice, as it means no runtime overhead and no UB to users of said struct

Reply:
> Padding bits exist at runtime, that's just a fact. They're bits that _pad_ the surrounding struct to the correct size and/or alignment, as you said. These bits have a value of _some_ kind at runtime, and that value needs to be defined in some way, be it undefined value, unspecified value or a specified value.

Totally agree. But this has nothing to do with UnnamedStructElement. While UnnamedStructElement affects the layout of the struct and can thereby cause padding bits to appear, padding bits also exist in structs without UnnamedStructElements. So what you are proposing is that the behavior of padding bits should be specified by this RFC for repr(C) structs with bit-fields, the behavior of padding bits in any other context stays unspecified. (Unless the behavior of padding bits is already specified in general somewhere else in which case it is unnecessary to specify it again in this special case.)

Reply:
The defined-ness of any other kind of padding is irrelevant to this pr except for the lessons we can learn from it. Bitfields are very low-level types and so a certain level of tomfoolery (if you will) is to be expected from them, the exact kinds of tomfoolery that are mercilessly preyed upon by the undefined behavior induced by padding with undefined contents. In light of that and given the position of defining an entirely new data type that's free from any prior attachment to the padding woes of `repr(Rust)` or anything else, we can simply give the padding bits an unspecified value and make it so that there's no UB involved with them whatsoever. This pr shouldn't and can't solve or deal with the problems of reading padding bits from other structs, nor does it have to at all since this is an entirely new data type.

Reply:
To be clear, are you proposing that we should specify the behavior of padding bits in structs with bit-fields while not addressing the behavior of padding bits in any other kind of struct?

If so then I've misunderstood you so far and I can imagine doing that. But we'd first have to consult the experts that have discussed these topics for several years.

Reply:
Yes, padding of other structs is irrelevant to the bitfield proposal. This rfc's goal isn't to solve the problem of padding, but it has the choice of either *adding* to it or making it a non-issue, and I think it should make it a non-issue by giving the padding bits of bitfields an unspecified value

Reply:
I imagine that the immediate counter will be that this would unreasonably restrict future possibilities for defining the behavior of padding bits in general.

Reply:
I don't see how it's restrictive at all, the value is unspecified and hence can't be relied on for anything. Feasibly any kind of change could be made except for making the padding undefined, which shouldn't happen and I don't see anyone from Rust advocating for such

Reply:
Anyway, if you can build consensus around defining the behavior of padding bits in structs with bit-fields, then I'll add it to the RFC. But I'll not do this myself because I think that it will take a significant amount of time to research this issue.

Reply:
cc @Lokathor @joshtriplett

Reply:
Kixiron said what I was trying to say.

Essentially, if the spare bits are defined to be anything other than "initialized but unspecified" then I can make a *more safe* version of bitfields, by hand, today, than this RFC would allow for.

Reply:
I think I see why this was initially confusing.

It'd be conceptually straightforward to interpret the unnamed fields as "padding", except for the `bitfield(0)` case. But then it's necessary to specify the behavior of those fields.

You're proposing, instead, that unnamed "fields" (including `bitfield(0)` fields) don't define padding themselves; rather, they define the positioning of bitfields, and the positioning of bitfields within a larger structure defines padding. And that padding behaves like any other padding.

I think this explanation could use further in-RFC clarification to make the approach more obvious, and to separate the concept of "unnamed 'fields' defining where fields are placed" from "padding defined by where fields are placed". But given some additional explanation in that area, I don't object to the idea of making this RFC entirely orthogonal to any definition of padding..

Reply:
> You're proposing, instead, that unnamed "fields" (including `bitfield(0)` fields) don't define padding themselves; rather, they define the positioning of bitfields, and the positioning of bitfields within a larger structure defines padding.

No I think that's too narrow. Unnamed "fields" also affect the positioning of regular fields and the overall alignment. They are layout modifiers like `repr(align)` except that they are written inside the struct body.

Reply:
And, reading further into the updated description, I think the new approach of not calling them "fields" provides the necessary clarity. (The distinction of "bit-field fields" is perhaps a little awkward, but nonetheless clear.)